pystac_client/collection_client.py

```diff
@@ -70,6 +71,15 @@ def from_dict(
         setattr(result, "modifier", modifier)
         return result
 
+    def set_root(self, root: Optional[pystac.Catalog]) -> None:
+        # hook in to set_root and use it for setting _stac_io
+        super().set_root(root=root)
+        if root is not None and root._stac_io is not None:
+            if not isinstance(root._stac_io, StacApiIO):
+                raise ValueError("Root should be a Client object")
+            else:
+                self._stac_io = root._stac_io
```

Comment:
There's other places where we rely on our root to hold our `_stac_io` (e.g. https://github.com/stac-utils/pystac/blob/424e7482536ae845f6cc96ae56de243a084a6b0a/pystac/stac_object.py#L410-L418). Is there a reason we flip that pattern here?